Image Upload: loading state UI and integration

jaclync opened the pull request:
"UI (Product image loading state) and integration with MediaStore" part of #1713 

## Summary

This PR created the UI for Product image loading state, and integrated the UI with image upload functionality.

## Changes

- Created `CircleSpinnerView` to display an animatable circle spinner as in the design spec
- Created `InProgressProductImageCollectionViewCell` cell when an image is still being processed/uploaded
- Created an enum `ProductImageStatus` that indicates the status of a Product image with 2 cases for now: remote image, or uploading an asset
- Created `ProductImagesService` that encapsulates the implementation of uploading an image asset
- In `ProductImagesViewController` and `ProductImagesCollectionViewController`, updated its data source to an array of `ProductImageStatus` instead of `ProductImage` and displayed the cell and image based on the status enum
- In `ProductImagesViewController`, uploaded a media asset with `ProductImagesService` and updated the corresponding cell upon an error or remote `ProductImage`
  - Since the order of the image statuses could change when we implement reordering, it searches the first index path matching the asset

## Testing

Testing on a device is recommeded so that the camera is enabled

- Launch the app
- Go to Products tab
- Tap on a simple Product
- Tap on the + UI in the images header
- Tap "Add Photos"
- Tap "Choose from device"
- Pick a few photos from the photo library
- Tap "Add #" in the navigation bar --> should see the correct number of images added to the images UI with loading state
- Wait for a bit, and feel free to play around like tapping "Done", navigating back, tapping on another image and deleting it --> the image(s) should be uploaded given network connection
- Repeat the above steps with the camera from "Add Photos" action sheet

## Example screenshots

loading UI | image upload
-- | --
<img src="https://user-images.githubusercontent.com/1945542/73436170-f9911380-4384-11ea-9209-6f3354ba2201.PNG" width="600" /> | ![image_upload_v1](https://user-images.githubusercontent.com/1945542/73438649-54c50500-4389-11ea-89b3-e4b0a0a2d297.gif)


Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

pmusolino commented on WooCommerce/Classes/ViewRelated/ReusableViews/CircleSpinnerView.swift:
```diff
@@ -0,0 +1,128 @@
+
+/// An animatable circular spinner.
+///
+@IBDesignable
```
I think that we need to remove `@IBDesignable` here since we will allow configuring this View from Xib or Storyboard, it is something that we should avoid since we try to configure everything possible from code (and maybe it can create confusion, making the view configuration "uneven").

pmusolino commented on WooCommerce/Classes/ViewRelated/ReusableViews/CircleSpinnerView.swift:
```diff
@@ -0,0 +1,128 @@
+
+    // MARK: Configurable properties
+
+    @IBInspectable
```
Like `@IBDesignable` 👍

astralbodies commented:
@jaclync @pmusolino I did something unconventional and added feature flag support in your PR to get Products M2 work to be visible for the installable builds. Don't be mad 😄

jaclync commented:
> For my first upload, I tried to upload 2 images and the app crashed, but it worked when I tried the second time

@kyleaparker do you remember the steps before the first crash?

jaclync commented on WooCommerce/Classes/ViewRelated/Products/Media/ProductImagesCollectionViewController.swift:
```diff
@@ -74,17 +97,37 @@ extension ProductImagesCollectionViewController {
                                                             cell.imageView.contentMode = .center
```
yeaa great catch! in [this commit](https://github.com/woocommerce/woocommerce-ios/pull/1800/commits/e70fb80c0d8d34cc69a4942b7f38c3ff0a48ec64) I set the content mode to `center` first, and then updated it to `scaleAspectFit` after the image is successfully loaded

jaclync commented:
>  If I start to upload some images and I press save, I lost all the images that I was uploading, but from the website, I'm able to see that these images are uploaded correctly in the background. What is the expected behavior in this particular case? 🤔

Since we removed the "Save changes" option in the action sheet, now when tapping "Done" button while an image is still being uploaded, we show an action sheet that indicates some outstanding changes:

<img src="https://user-images.githubusercontent.com/1945542/73813009-b79b1e00-4819-11ea-85fd-98a1677271ce.png" width="320" />

I'll also clarify the expected behavior with the team, will update!

> The placeholder of the image is big (fit the cell size) (see the image below). I left a comment with a possible fix.
Feel free to manage them in a separated PR :)

I updated it in https://github.com/woocommerce/woocommerce-ios/pull/1800/commits/e70fb80c0d8d34cc69a4942b7f38c3ff0a48ec64, see if it works for you now :)

kyleaparker commented:
>@kyleaparker do you remember the steps before the first crash?

Yep!
1. Opened Photos screen
2. Clicked Add Photos
3. Clicked Choose from device
4. Chose two pictures

jaclync commented:
@pmusolino Based on p99K0U-2mN-p2#comment-4732, we will keep uploading the images to WP media library in any case. If the user taps "Done" while some images are still being uploaded, we will keep track of these in-progress images (to be implemented in another subtask, probably going to implement some kind of observation). In https://github.com/woocommerce/woocommerce-ios/pull/1800/commits/bda25a643c3de859dc38f7fb5603d72dc63e00a9, I updated to only show the Discard Changes action sheet when some images are deleted while navigating away. Ready for another look!

@kyleaparker I installed the installable build but I wasn't to reproduce the crash and the crashes aren't recorded anywhere now (I made a request to #platform9p2) 😞I will keep an out for it!

jaclync commented:
merging this PR now, thanks for testing/reviewing Kylea & Paolo!
